File: 0005. Longest Palindromic Substring.py
from typing import *

class Solution:
    def longestPalindrome(self, s: str) -> str:
        def palindrome_boundaries(left, right, str):
            while left >= 0 and right < len(str) and str[left] == str[right]:
                left -= 1
                right += 1

            return left+1, right-1

        n = len(s)
        ans = ''

        for i in range(n):
            first_left, first_right = palindrome_boundaries(i, i, s)

            if first_right - first_left + 1 > len(ans):
                ans = s[first_left:first_right+1]

            if i+1 < n:
                second_left, second_right = palindrome_boundaries(i, i+1, s)

                if second_right - second_left + 1 > len(ans):
                    ans = s[second_left:second_right+1]

        return ans

File: js/0005. Longest Palindromic Substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    var n = s.length, ans = '';
    const LEFT = 0, RIGHT = 1;
    
    for (let i = 0; i < n; i++) {
        let bounds = getBoundaries(s, i, i);
        if (bounds[RIGHT] - bounds[LEFT] + 1 > ans.length) {
            ans = s.slice(bounds[LEFT], bounds[RIGHT]+1);
        }
        
        if (i+1 < n) {
            let bounds = getBoundaries(s, i, i+1);
            if (bounds[RIGHT] - bounds[LEFT] + 1 > ans.length) {
                ans = s.slice(bounds[LEFT], bounds[RIGHT]+1);
            }
        }
    }
    
    return ans;
};

var getBoundaries = function(s, left, right) {
    while (left >= 0 && right < s.length && s[left] == s[right]) {
        left -= 1;
        right += 1;
    }
    
    return [left+1, right-1];
};
